include/ringmesh/mesh/cartesian_grid.h

```diff
@@ -0,0 +1,190 @@
+ *
+ *     http://www.ring-team.org
+ *
+ *     RING Project
+ *     Ecole Nationale Superieure de Geologie - GeoRessources
+ *     2 Rue du Doyen Marcel Roubault - TSA 70605
+ *     54518 VANDOEUVRE-LES-NANCY
+ *     FRANCE
+ */
+
+#pragma once
+
+#include <ringmesh/basic/common.h>
+
+//#include <memory>
+
+//#include <geogram/mesh/mesh.h>
+//#include <geogram/mesh/mesh_io.h>
+
+//#include <ringmesh/geogram_extension/geogram_extension.h>
+
+#include <ringmesh/basic/frame.h>
+//#include <ringmesh/mesh/mesh.h>
+//#include <ringmesh/mesh/mesh_index.h>
+
+namespace RINGMesh
+{
+    FORWARD_DECLARATION_DIMENSION_CLASS( CartesianGridBuilder );
+}
+
+/*!
+ * @file Generic class for creating cartesian grids
+ * @author Melchior Schuh-Senlis
+ */
+
+namespace RINGMesh
+{
+    template < index_t DIMENSION >
+    class RINGMESH_API CartesianGrid
+    {
+        friend class CartesianGridBuilder< DIMENSION >;
+
+    public:
+        CartesianGrid( intvecn< DIMENSION > nb_cells,
+            index_t nb_total_cells,
+            ReferenceFrame< DIMENSION > vec_cartesian_axis )
+            : nb_cells_( nb_cells ), cartesian_frame_( vec_cartesian_axis )
+        {
+            nb_total_cells_ = 1;
+            for( auto i : RINGMesh::range( DIMENSION ) )
+            {
+                nb_total_cells_ *= nb_cells;
+            }
+        }
+
+        static MeshType type_name_static()
+        {
+            return 'CartesianGrid';
+        }
+        void save_mesh( const std::string& filename ) const
+        {
+            // TODO
+        }
+        void print_mesh_bounded_attributes() const
+        {
+            // TODO
+        }
+        //		GEO::AttributesManager& vertex_attribute_manager() const
+        // override
+        //		{
+        //			return mesh_->vertices.attributes();
+        //		}
+        MeshType type_name() const override
+        {
+            return type_name_static();
+        }
+        //		static std::string default_extension_static()
+        //		{
+        //			return "geogram";
+        //		}
+        //		std::string default_extension() const override
+        //		{
+        //			return default_extension_static();
+        //		}
+        //		index_t nb_vertices() const override
+        //		{
+        //			return mesh_->vertices.nb();
+        //		}
+
+    private:
+        /**
+         * \brief Forbids copy.
+         * \details This is to make sure that client code does
+         *   not unintentionally copy a Mesh (for
+         *   instance by passing it by-value to a function).
+         *   Use copy() instead.
+         */
+        CartesianGrid( const CartesianGrid& rhs );
+
+        /**
+         * \brief Forbids copy.
+         * \details This is to make sure that client code does
+         *   not unintentionally copies a Mesh (for
+         *   instance by passing it by-value to a function).
+         *   Use copy() instead.
+         */
+        const CartesianGrid& operator=( const CartesianGrid& rhs );
```

Comment:
A macro exists in RINGMesh to delete these functions : ringmesh_disable_copy( Class )

Reply:
We should use them in VolumeMesh too then (copied some stuff from there).